### DISCUSSION TITLE
SSR and high memory consumption by react-query cache

### DISCUSSION THREAD

**Discussion by meskill** [TanStack/query#3284] [Q&A] [answered]
I faced the issue with high memory consumption on nodejs app with SSR and react-query@3.34.14.

Profiling the memory leads to the react-query cache internals. As I understand, a cache for the server works as explained in [docs](https://react-query.tanstack.com/guides/caching) with isolation for every request. 

The problem is that cache for every request stays in server memory for the `cacheTime` and it is not collected by GC for that time. 

Explicit call to [queryClient.clear()](https://react-query.tanstack.com/reference/QueryClient#queryclientclear) after sending the response fixes the problem, but it isn't mentioned in [docs for ssr](https://react-query.tanstack.com/guides/ssr#on-the-server)

Questions: Is this intentional how the cache works on server? Is call to `queryClient.clear()` on SSR is right thing to do?

**Comment by TkDodo** [comment, marked as answer]
are you creating the queryClient inside of getServerSideProps / getStaticProps?

```
export async function getStaticProps() {
   const queryClient = new QueryClient()
 
   await queryClient.prefetchQuery('posts', getPosts)
 
   return {
     props: {
       dehydratedState: dehydrate(queryClient),
     },
   }
 }
```

but yeah the running timers might actually stop it from being garbage collected by node. Have you tried setting a smaller cacheTime, like 10 seconds? 0 won't work because then it could be empty by the time you call `dehydrate`...

> Profiling the memory leads to the react-query cache internals.

Can you share the profiling results maybe?

**Reply by meskill** [reply]
We have a custom implementation for SSR that creates queryClient for every request.

I cannot share actual profiling but react-query definitely prevents data from being collected by GC for the `cacheTime`. And it is critical for our app as it holds a big DI object with every request-related data.

I think lowering `cacheTime` might make things better, but currently, I've added `queryClient.clear()` call after request handled to force cache cleanup.

**Reply by TkDodo** [reply]
yes that should also work. In NextJs, if you add it in `gerServerSideProps`, it would also create a new client in every request I think. So maybe adding at least a caveat to the SSR docs would be good I think. Would you like to do that?

**Reply by meskill** [reply]
Please, take a look https://github.com/tannerlinsley/react-query/pull/3290

**Comment by phaethon5882** [comment]
@TkDodo @meskill 
I have one question!
When exactly should you call queryClient.clear() ?
Can you show an example?

**Reply by meskill** [reply]
You can find example in the [doc](https://tanstack.com/query/v4/docs/guides/ssr#on-the-server)

**Reply by phaethon5882** [reply]
thx @meskill!
I'm using nextjs, so I'll have to find another way.
If I figure out how, I'll leave a comment.

**Comment by rafakwolf** [comment]
Same question as @phaethon5882  ...
does it solve?

```
export async function getServerSideProps() {
  const queryClient = new QueryClient();
  let dehydratedState = {};
  try {
   //  await my endpoint request
   dehydratedState = dehydrate(queryClient);
   queryClient.clear();
  } catch (err) {
    // error treatment
  }

  return {
    props: {
      dehydratedState,
    }
  };
}
```

Thanks !

**Reply by TkDodo** [reply]
we changed how that works in v4: https://tanstack.com/query/v4/docs/guides/migrating-to-react-query-4#no-default-manual-garbage-collection-server-side

**Reply by rafakwolf** [reply]
@TkDodo thank you for the quick answer ...
We're using v4, but still seeing high memory consumption on the server side ...
Does the change you mentioned mean that manually cleaning **queryClient** will do nothing?

**Reply by TkDodo** [reply]
`cacheTime` being set to `Infinity` on the server means that there are no timers running that could stop the cache from being garbage collected by node. That was what previously caused the high consumption. If you made many requests, each time a new QueryClient would be created that would sit around for 5 minutes. Now, the QueryClient should just be garbage collected like any other unused class / variable.

So yeah, you shouldn't need to clear the query cache anymore.

**Reply by c01inQvQ** [reply]
Hi, @TkDodo

We're currently using NextJS and React-Query v3 with the default 5mins `cacheTime`. 
Does the code snippet shared by @rafakwolf solve the high memory consumption issue?

Thanks !

**Reply by TkDodo** [reply]
yes. or, set cacheTime to Infinity.

**Reply by c01inQvQ** [reply]
@TkDodo Thanks for your quick anwser !

**Reply by gavinthomas-valtech** [reply]
>e're currently using NextJS and React-Query v3 with the default 5mins cacheTime.
Does the code snippet shared by @rafakwolf solve the high memory consumption issue?

Are we saying v3 suffers from high memory consumption on the server if you don't clear the queryClient when using SSR? Was there a change in v4?

**Reply by gavinthomas-valtech** [reply]
Just spotted this

>In v3, React Query would cache query results for a default of 5 minutes, then manually garbage collect that data. This default was applied to server-side React Query as well.

This lead to high memory consumption and hanging processes waiting for this manual garbage collection to complete. In v4, by default the server-side cacheTime is now set to Infinity effectively disabling manual garbage collection (the NodeJS process will clear everything once a request is complete).

This change only impacts users of server-side React Query, such as with Next.js. If you are setting a cacheTime manually this will not impact you (although you may want to mirror behavior).

How do i set Infinity cacheTime on the server but 5 minutes client side?

**Reply by TkDodo** [reply]
> Are we saying v3 suffers from high memory consumption on the server if you don't clear the queryClient when using SSR? Was there a change in v4?

yes, in v4, cacheTime defaults to Infinity on the server. I mean, we do write migration docs for these cases: https://tanstack.com/query/v4/docs/react/guides/migrating-to-react-query-4#no-default-manual-garbage-collection-server-side

> How do i set Infinity cacheTime on the server but 5 minutes client side?

you have a different QueryClient on the server and on the client, so you can give them different options.

**Comment by magicqiang** [comment]
I think this new v4 update for setting cacheTime to be Infinity or query.clear() only works if not passing in any custom cacheTime

The memory leak issue will still happen if useQuery is used inside components with custom cacheTime while being rendered in ssr.

I tried to have a root clean up method where we pass queryClient to QueryClientProvider

```ts
useEffect(() => {
return  () => {
  queryClient.clear()
 }
}, [])
```

but it still didn't solve the problem. I then manually change where we customize cacheTime to be 

```ts
cacheTime: typeof window !== 'undefined' ? Infinity:  YOUR_OWN_CUSTOM_TIME
```
which seems working

@TkDodo do you think it makes sense that react-query enforce cacheTime to be Infinity by default on server side regardless of what custom cache time user pass in?

**Reply by TkDodo** [reply]
> I tried to have a root clean up method where we pass queryClient to QueryClientProvider

Effects don't run on the server, so it's not surprising that this doesn't work.

> @TkDodo do you think it makes sense that react-query enforce cacheTime to be Infinity by default on server side regardless of what custom cache time user pass in?

No, I don't think so, because it depends on how you use react-query on the server. If you use it for prefetching, where you create a new, isolated client per request, then `cacheTime` will start a timer for the queries in your cache (because they are unused), which means node can't garbage collect it. It's not per se a "memory leak", it's just still used and therefore not cleaned up. In those cases, you want `cacheTime: 0`.

However, you might want to use react-query on the server for server-side, in-memory caching, which is a different use-case. There, setting a `cacheTime` might make sense - you'd want your unused queries to be garbage collected at some point. You will also likely only create one client, so it won't be noticeable that the timers still run.

We've optimized towards the former usage, but we can't ignore the other use-case, so we can't enforce zero cacheTime at all costs.
